packages/open-next/src/adapters/routing.ts

```diff
@@ -0,0 +1,109 @@
+import { InternalEvent } from './event-mapper.js';
+import { debug } from './logger.js';
+import {
+  RedirectDefinition,
+  RewriteDefinition,
+  RewriteMatcher,
+} from './next-types.js';
+import { compile, match } from 'path-to-regexp';
+
+const redirectMatcher =
+  (
+    headers: Record<string, string>,
+    cookies: Record<string, string>,
+    query: Record<string, string | string[]>
+  ) =>
+  (redirect: RewriteMatcher) => {
+    switch (redirect.type) {
+      case 'header':
+        return (
+          headers?.[redirect.key.toLowerCase()] &&
+          new RegExp(redirect.value ?? '').test(
+            headers[redirect.key.toLowerCase()] ?? ''
+          )
+        );
+      case 'cookie':
+        return (
+          cookies?.[redirect.key] &&
+          new RegExp(redirect.value ?? '').test(cookies[redirect.key] ?? '')
+        );
+      case 'query':
+        return query[redirect.key] && Array.isArray(redirect.value)
+          ? redirect.value.reduce(
+              (prev, current) =>
+                prev || new RegExp(current).test(query[redirect.key] as string),
+              false
+            )
+          : new RegExp(redirect.value ?? '').test(
+              (query[redirect.key] as string | undefined) ?? ''
+            );
+      case 'host':
+        return (
+          headers?.host && new RegExp(redirect.value ?? '').test(headers.host)
+        );
+      default:
+        return false;
+    }
+  };
+
+export function handleRewrites<T extends RewriteDefinition>(
+  internalEvent: InternalEvent,
+  rewrites: T[]
+) {
+  const { rawPath, headers, query, cookies } = internalEvent;
+  const matcher = redirectMatcher(headers, cookies, query);
+  const rewrite = rewrites.find(
+    (route) =>
+      new RegExp(route.regex).test(rawPath) &&
+      (route.has
+        ? route.has.reduce((acc, cur) => {
+            if (acc === false) return false;
+            return matcher(cur);
+          }, true)
+        : true) &&
+      (route.missing
+        ? route.missing.reduce((acc, cur) => {
+            if (acc === false) return false;
+            return !matcher(cur);
+          }, true)
+        : true)
+  );
+
+  const urlQueryString = new URLSearchParams(query).toString();
+  let rewrittenUrl = rawPath;
+  if (rewrite) {
+    const toDestination = compile(rewrite?.destination ?? '');
+    const fromSource = match(rewrite?.source ?? '');
+    const _match = fromSource(rawPath);
+    if (_match) {
+      const { params } = _match;
+      const isUsingParams = Object.keys(params).length > 0;
+      if (isUsingParams) {
+        rewrittenUrl = toDestination(params);
+      }
+      rewrittenUrl = rewrite.destination;
```

Comment:
This should be inside an `else` statement, otherwise the `toDestination` gets overriden.

Reply:
Ho come on, how could i forget that..
Ok at least this way it makes more sense to me, i couldn't understand how your fix could resolve the issue, i'm pretty sure we need the `(.)` in the final url.
I've just tried something and it seems to work

I created these 2 functions 
```typescript
export function escapeRegex(str: string) {
  let path = str.replace(/\(\.\)/g, '%1');

  path = path.replace(/\(\.{2}\)/g, '%2');

  path = path.replace(/\(\.{3}\)/g, '%3');

  return path;
}

export function unescapeRegex(str: string) {
  let path = str.replace(/%1/g, '(.)');

  path = path.replace(/%2/g, '(..)');

  path = path.replace(/%3/g, '(...)');

  return path;
}
```
And for the rewrite part 
```typescript
const toDestination = compile(
      escapeRegex(rewrite?.destination ?? '') ?? ''
    );
    const fromSource = match(escapeRegex(rewrite?.source) ?? '');
    const _match = fromSource(rawPath);
    if (_match) {
      const { params } = _match;
      const isUsingParams = Object.keys(params).length > 0;
      if (isUsingParams) {
        rewrittenUrl = unescapeRegex(toDestination(params));
        console.log('rewrittenUrl1', rewrittenUrl);
      } else {
        rewrittenUrl = rewrite.destination;
      }
      console.log('rewrittenUrl2', rewrittenUrl);
```

Of course those escape character are pretty bad, but for testing it's ok, i need to figure out some escape that will not mess things up

Reply:
I have an extremely ugly hack to get the correct params from the modal:

In routing.ts:
```
  let params: { [key: string]: string } = {};

  if (rewrite) {
    const fromSource = match(resolvePaths(rewrite?.source) ?? "");
    const _match = fromSource(rawPath);
    if (_match) {
      params = _match.params as any;
      rewrittenUrl = rewrite.destination;
    }
  }

  return {
    rawPath: rewrittenUrl,
    url: `${rewrittenUrl}${urlQueryString ? `?${urlQueryString}` : ""}`,
    params,
    __rewrite: rewrite,
  };
  ```
In server-adapter:
```
  const { rawPath, url, params } = handleRewrites(
    internalEvent,
    routesManifest.rewrites
  );
....
  const body = ServerResponse.body(res)
    .toString(encoding)
    .replace(/%3A([^"]+)/g, (_, match: string) => {
      return params[match] || `%3A${match}`;
    });
```

But there's gotta be a correct way to do this...

Reply:
Cool, let me give that a go.

Reply:
Nice! Modal interception + client parameters works now :)

Reply:
Would `fromSource` response ever need to be unescapeRegex'd?

Reply:
Yes otherwise `next-server` will not find the correct interception route path.
I'll change the escape character to something like`_µ1_`, it should be safe enough so that user will not have these in the url